# IDVE Assignment 2
## NBA 2022-23 Season Statistics Analysis

**Student Name:** [Your Name]
**Student ID:** [Your ID]
**Date:** [Date]


## Setup and Imports

Import all necessary libraries for data analysis and visualization.

In [1]:
# Data manipulation and analysis
import pandas as pd
import numpy as np

# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Interactive visualizations
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Statistical analysis
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette("husl")

# Ignore warnings
import warnings
warnings.filterwarnings('ignore')

print("All libraries imported successfully!")

All libraries imported successfully!


## 1. Data Loading and Initial Exploration

Load the NBA 2022-23 statistics dataset and perform initial exploration.

In [2]:
# Load the dataset
df = pd.read_csv('nba_2022-23_stats.csv')

print(f"Dataset shape: {df.shape}")
print(f"Number of players: {df.shape[0]}")
print(f"Number of features: {df.shape[1]}")

Dataset shape: (467, 52)
Number of players: 467
Number of features: 52


In [3]:
# Display first few rows
df.head(10)

Unnamed: 0            Player Name    Salary Position  Age     Team  GP  GS  \
0           0          Stephen Curry  48070014       PG   34      GSW  56  56   
1           1              John Wall  47345760       PG   32      LAC  34   3   
2           2      Russell Westbrook  47080179       PG   34  LAL/LAC  73  24   
3           3           LeBron James  44474988       PF   38      LAL  55  54   
4           4           Kevin Durant  44119845       PF   34  BRK/PHO  47  47   
5           5           Bradley Beal  43279250       SG   29      WAS  50  50   
6           6          Kawhi Leonard  42492492       SF   31      LAC  52  50   
7           7            Paul George  42492492       SF   32      LAC  56  56   
8           8  Giannis Antetokounmpo  42492492       PF   28      MIL  63  63   
9           9         Damian Lillard  42492492       PG   32      POR  58  58   

     MP    FG   FGA    FG%    3P   3PA    3P%    2P   2PA    2P%   eFG%   FT  \
0  34.7  10.0  20.2  0.493   4.9  11.4  0.427   5.1   8.8  0.579  0.614  4.6   
1  22.2   4.1   9.9  0.408   1.0   3.2  0.303   3.1   6.7  0.459  0.457  2.3   
2  29.1   5.9  13.6  0.436   1.2   3.9  0.311   4.7   9.7  0.487  0.481  2.8   
3  35.5  11.1  22.2  0.500   2.2   6.9  0.321   8.9  15.3  0.580  0.549  4.6   
4  35.6  10.3  18.3  0.560   2.0   4.9  0.404   8.3  13.4  0.617  0.614  6.5   
5  33.5   8.9  17.6  0.506  1.s6   4.4  0.365   7.3  13.2  0.552  0.551  3.8   
6  33.6   8.6  16.8  0.512   2.0   4.8  0.416   6.6  11.9  0.551  0.572  4.7   
7  34.6   8.2  17.9  0.457   2.8   7.6  0.371   5.4  10.3  0.521  0.536  4.6   
8  32.1  11.2  20.3  0.553   0.7   2.7  0.275  10.5  17.6  0.596  0.572  7.9   
9  36.3   9.6  20.7  0.463   4.2  11.3  0.371   5.4   9.4  0.574  0.564  8.8   

    FTA    FT%  ORB  DRB   TRB  AST  STL  BLK  TOV   PF   PTS  Total Minutes  \
0   5.0  0.915  0.7  5.4   6.1  6.3  0.9  0.4  3.2  2.1  29.4           1941   
1   3.3  0.681  0.4  2.3   2.7  5.2  0.8  0.4  2.4  1.7  11.4            755   
2   4.3  0.656  1.2  4.6   5.8  7.5  1.0  0.5  3.5  2.2  15.9           2126   
3   5.9  0.768  1.2  7.1   8.3  6.8  0.9  0.6  3.2  1.6  28.9           1954   
4   7.1  0.919  0.4  6.3   6.7  5.0  0.7  1.4  3.3  2.1  29.1           1672   
5   4.6  0.842  0.8  3.1   3.9  5.4  0.9  0.7  2.9  2.1  23.2           1673   
6   5.4  0.871  1.1  5.4   6.5  3.9  1.4  0.5  1.7  1.6  23.8           1748   
7   5.3  0.871  0.8  5.3   6.1  5.1  1.5  0.4  3.1  2.8  23.8           1939   
8  12.3  0.645  2.2  9.6  11.8  5.7  0.8  0.8  3.9  3.1  31.1           2024   
9   9.6  0.914  0.8  4.0   4.8  7.3  0.9  0.3  3.3  1.9  32.2           2107   

    PER    TS%   3PAr    FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  \
0  24.1  0.656  0.564  0.248   2.3  16.8   9.7  30.0   1.3   0.9  12.5  31.0   
1  13.6  0.498  0.322  0.334   2.1  11.4   6.8  35.3   1.8   1.4  17.1  27.0   
2  16.1  0.513  0.289  0.317   4.7  16.5  10.8  38.6   1.7   1.3  18.4  27.7   
3  23.9  0.583  0.309  0.268   3.7  20.8  12.5  33.5   1.2   1.4  11.6  33.3   
4  25.9  0.677  0.267  0.387   1.2  19.5  10.5  24.5   1.0   3.4  13.4  30.7   
5  19.7  0.593  0.249  0.260   2.8   9.9   6.5  26.6   1.3   1.7  12.9  29.2   
6  23.9  0.623  0.287  0.320   3.7  17.8  10.8  19.6   2.0   1.4   8.1  27.0   
7  19.6  0.588  0.424  0.293   2.6  17.0   9.9  24.2   2.1   0.9  13.4  29.5   
8  29.0  0.605  0.134  0.604   7.3  30.0  19.1  33.2   1.2   2.1  13.2  38.8   
9  26.7  0.645  0.547  0.464   2.4  12.8   7.6  35.0   1.2   0.8  11.7  33.8   

   OWS  DWS   WS  WS/48  OBPM  DBPM  BPM  VORP  
0  5.8  2.0  7.8  0.192   7.5   0.1  7.5   4.7  
1 -0.4  0.7  0.3  0.020  -0.8  -0.4 -1.2   0.1  
2 -0.6  2.6  1.9  0.044   0.3  -0.1  0.2   1.2  
3  3.2  2.4  5.6  0.138   5.5   0.6  6.1   4.0  
4  4.7  2.1  6.8  0.194   6.0   1.2  7.1   3.9  
5  2.2  1.2  3.4  0.099   2.9  -1.2  1.8   1.6  
6  4.9  2.2  7.1  0.194   5.1   0.9  6.1   3.5  
7  2.3  2.3  4.6  0.114   2.4   0.3  2.8   2.3  
8  4.9  3.7  8.6  0.204   5

In [4]:
# Display dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 467 entries, 0 to 466
Data columns (total 52 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     467 non-null    int64  
 1   Player Name    467 non-null    object 
 2   Salary         467 non-null    int64  
 3   Position       467 non-null    object 
 4   Age            467 non-null    int64  
 5   Team           467 non-null    object 
 6   GP             467 non-null    int64  
 7   GS             467 non-null    int64  
 8   MP             467 non-null    float64
 9   FG             467 non-null    float64
 10  FGA            467 non-null    float64
 11  FG%            466 non-null    float64
 12  3P             467 non-null    object 
 13  3PA            467 non-null    float64
 14  3P%            454 non-null    float64
 15  2P             467 non-null    float64
 16  2PA            467 non-null    float64
 17  2P%            463 non-null    float64
 18  eFG%      

In [5]:
# Display statistical summary
df.describe()

Unnamed: 0        Salary         Age          GP          GS  \
count  467.000000  4.670000e+02  467.000000  467.000000  467.000000   
mean   233.000000  8.416599e+06   25.820128   48.233405   22.650964   
std    134.955548  1.070812e+07    4.275113   24.807740   27.094577   
min      0.000000  5.849000e+03   19.000000    1.000000    0.000000   
25%    116.500000  1.782621e+06   23.000000   31.000000    1.000000   
50%    233.000000  3.722040e+06   25.000000   55.000000    8.000000   
75%    349.500000  1.063354e+07   29.000000   68.500000   45.500000   
max    466.000000  4.807001e+07   42.000000   83.000000   83.000000   

               MP          FG         FGA         FG%         3PA         3P%  \
count  467.000000  467.000000  467.000000  466.000000  467.000000  454.000000   
mean    19.871306    3.351392    7.117773    0.465009    2.792719    0.325090   
std      9.548684    2.457836    5.020700    0.109547    2.260794    0.131624   
min      1.800000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%     12.500000    1.600000    3.300000    0.417000    1.000000    0.288000   
50%     19.200000    2.700000    5.800000    0.455000    2.400000    0.344500   
75%     28.300000    4.300000    9.400000    0.507500    4.150000    0.387000   
max     41.000000   11.200000   22.200000    1.000000   11.400000    1.000000   

               2P         2PA         2P%        eFG%          FT         FTA  \
count  467.000000  467.000000  463.000000  466.000000  467.000000  467.000000   
mean     2.356531    4.325268    0.533464    0.534266    1.435546    1.848394   
std      1.989397    3.571280    0.136046    0.104584    1.568500    1.896352   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.900000    1.800000    0.487000    0.493750    0.500000    0.600000   
50%      1.700000    3.200000    0.537000    0.537500    0.900000    1.200000   
75%      3.300000    5.850000    0.594500    0.578000    1.850000    2.350000   
max     10.500000   17.800000    1.000000    1.000000   10.000000   12.300000   

              FT%         ORB         DRB         TRB         AST         STL  \
count  444.000000  467.000000  467.000000  467.000000  467.000000  467.000000   
mean     0.753489    0.866809    2.662099    3.528051    2.107709    0.610493   
std      0.148659    0.746976    1.707984    2.277160    1.957517    0.399083   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.691000    0.400000    1.400000    1.900000    0.800000    0.300000   
50%      0.769000    0.700000    2.300000    3.000000    1.400000    0.600000   
75%      0.844000    1.100000    3.500000    4.500000    2.900000    0.800000   
max      1.000000    5.100000    9.600000   12.500000   10.700000    3.000000   

              BLK         TOV          PF         PTS  Total Minutes  \
count  467.000000  467.000000  467.000000  467.000000     467.000000   
mean     0.379015    1.114133    1.702784    9.129979    1109.676660   
std      0.364387    0.830051    0.776957    6.906006     827.367432   
min      0.000000    0.000000    0.000000    0.000000       2.000000   
25%      0.100000    0.500000    1.200000    4.100000     341.000000   
50%      0.300000    0.900000    1.700000    7.100000     990.000000   
75%      0.500000    1.500000    2.200000   11.700000    1842.000000   
max      2.500000    4.100000    5.000000   33.100000    2963.000000   

              PER         TS%        3PAr         FTr        ORB%        DRB%  \
count  467.000000  466.000000  466.000000  466.000000  467.000000  467.000000   
mean    13.279657    0.563024    0.403655    0.249749    5.243469   14.884582   
std      6.230049    0.100741    0.219066    0.186855    4.332254    6.478093   
min    -20.900000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%     10.100000    0.524000    0.268250    0.143250    2.100000   10.700000   
50%     13.000000    0.567000    0.409000 

In [6]:
# Check for missing values
missing_values = df.isnull().sum()
print("Missing values per column:")
print(missing_values[missing_values > 0])

Missing values per column:
FG%      1
3P%     13
2P%      4
eFG%     1
FT%     23
TS%      1
3PAr     1
FTr      1
dtype: int64


In [7]:
# Display column names for reference
print("Available columns:")
for i, col in enumerate(df.columns, 1):
    print(f"{i:2d}. {col}")

Available columns:
 1. Unnamed: 0
 2. Player Name
 3. Salary
 4. Position
 5. Age
 6. Team
 7. GP
 8. GS
 9. MP
10. FG
11. FGA
12. FG%
13. 3P
14. 3PA
15. 3P%
16. 2P
17. 2PA
18. 2P%
19. eFG%
20. FT
21. FTA
22. FT%
23. ORB
24. DRB
25. TRB
26. AST
27. STL
28. BLK
29. TOV
30. PF
31. PTS
32. Total Minutes
33. PER
34. TS%
35. 3PAr
36. FTr
37. ORB%
38. DRB%
39. TRB%
40. AST%
41. STL%
42. BLK%
43. TOV%
44. USG%
45. OWS
46. DWS
47. WS
48. WS/48
49. OBPM
50. DBPM
51. BPM
52. VORP


## 2. Data Cleaning and Preprocessing

Handle missing values, data types, and prepare data for analysis.

This section documents all data cleaning operations performed on the NBA 2022-23 statistics dataset to ensure data quality and prepare it for analysis.

### Cleaning Steps Performed

#### 1. **Removed Redundant Index Column**
**Column:** `Unnamed: 0`

**Reason:** This column was an auto-generated index from the CSV file that duplicates the DataFrame's built-in index, providing no analytical value.

#### 2. **Fixed Data Type Issues in `3P` Column**
**Column:** `3P` (3-Point Field Goals Made)

**Problem:** The column contained string values with 's' suffix (e.g., '1.s6' instead of '1.6'), preventing numeric operations.

**Solution:** Removed 's' characters and converted to numeric type.

**Example:**
```
Before: '1.s6' (string)
After:  1.6 (float)
for Bradley Beal
```
#### 3. **Handled Missing Percentage Values by Filling with Zero**
**Columns:** `3P%`, `FT%`, `2P%`, `FG%`, `eFG%`, `TS%`, `3PAr`, `FTr`

**Problem:** Missing values in shooting percentage columns (23 missing in `FT%`, 13 in `3P%`, etc.)

**Solution:** Replaced all `NaN` values with `0`.

**Reasoning:**
- **For shooting percentages** (`3P%`, `FT%`, `2P%`, `FG%`): Missing values typically occur when a player had zero attempts. A 0% shooting percentage is mathematically correct when attempts = 0 (you can't make shots you don't take).
  
- **For efficiency metrics** (`eFG%`, `TS%`): These are calculated metrics. A value of 0 indicates no offensive production rather than "unknown," which is appropriate for players with minimal playing time.

- **For ratio metrics** (`3PAr`, `FTr`): These represent the rate of 3-point attempts or free throw attempts relative to field goal attempts. Zero indicates the player didn't utilize that aspect of offense.

**Example:**
```python
# Player with no 3-point attempts
Player: Steven Adams
3PA: 0.0 → 3P%: NaN → After cleaning: 0.0
Interpretation: Didn't attempt 3-pointers, so 0% is accurate

# Player with attempts but missing data (rare)
Player: Example Player
3PA: 5.2 → 3P%: NaN → After cleaning: 0.0
Interpretation: Potentially misleading - might need investigation
```

**Limitation:** This approach assumes missing percentages indicate zero attempts. If data is missing due to recording errors for players who did attempt shots, this could create inaccuracies.

#### 4. **Ensured Proper Data Types for All Numeric Columns**
**Columns:** All columns except `Player Name`, `Position`, `Team`

**Problem:** Some numeric columns were stored as objects (strings) instead of numeric types.

**Solution:** Converted all non-categorical columns to numeric types, with error handling for invalid values.

**Reason:** Numeric operations (mean, correlation, visualization) require proper numeric data types. String values would cause errors or be excluded from calculations.

**Example:**
```
Before: '34.7' (object/string)
After:  34.7 (float64)
```



In [8]:
# 1. Remove index column
df = df.drop('Unnamed: 0', axis=1)

# 2. Fix data type issues
df['3P'] = pd.to_numeric(df['3P'].astype(str).str.replace('s', ''), errors='coerce')

# 3. Handle missing percentage values
percentage_cols = ['3P%', 'FT%', '2P%', 'FG%', 'eFG%', 'TS%', '3PAr', 'FTr']
for col in percentage_cols:
    df[col] = df[col].fillna(0)

# 4. Ensure numeric columns are proper type
numeric_cols = df.select_dtypes(include=['object']).columns
for col in numeric_cols:
    if col not in ['Player Name', 'Position', 'Team']:
        df[col] = pd.to_numeric(df[col], errors='coerce')

# 5. Check final data quality
print("Missing values after cleaning:")
print(df.isnull().sum()[df.isnull().sum() > 0])

Missing values after cleaning:
Series([], dtype: int64)


In [9]:
# Check for players with very low playing time
print("=" * 60)
print("PLAYING TIME ANALYSIS")
print("=" * 60)
print(f"\nPlayers with < 100 total minutes: {(df['Total Minutes'] < 100).sum()}")
print(f"Players with < 10 games played: {(df['GP'] < 10).sum()}")
print(f"Players with < 5 MPG: {(df['MP'] < 5).sum()}")

# Show distribution of extreme values
print("\n" + "=" * 60)
print("EXTREME VALUES CHECK")
print("=" * 60)
print("\nPlayers with 100% shooting percentages (small samples):")
perfect_shooters = df[(df['FG%'] == 1.0) | (df['FT%'] == 1.0) | (df['3P%'] == 1.0)]
print(perfect_shooters[['Player Name', 'GP', 'Total Minutes', 'FG%', 'FT%', '3P%']].head(10))

print("\nPlayers with negative PER:")
print(df[df['PER'] < 0][['Player Name', 'GP', 'Total Minutes', 'PER']].head(10))

PLAYING TIME ANALYSIS

Players with < 100 total minutes: 64
Players with < 10 games played: 62
Players with < 5 MPG: 21

EXTREME VALUES CHECK

Players with 100% shooting percentages (small samples):
         Player Name  GP  Total Minutes    FG%    FT%    3P%
105      Danny Green  11            138  0.449  1.000  0.432
114   DeAndre Jordan  39            586  0.765  0.458  1.000
258  Markieff Morris  35            355  0.409  1.000  0.394
292      Bryn Forbes  25            268  0.361  1.000  0.304
312     Tony Bradley  12             33  0.500  1.000  0.600
349    Nick Richards  65           1217  0.629  0.749  1.000
362      Isaiah Todd   6             61  0.158  1.000  0.100
381      Noah Vonleh  23            171  0.458  1.000  0.250
383      Sam Merrill   5             59  0.409  1.000  0.278
384    Tyrese Martin  16             66  0.391  1.000  0.143

Players with negative PER:
          Player Name  GP  Total Minutes   PER
254      Joe Wieskamp   9             50  -0.4
356   Le

In [10]:
# Additional Cleaning - Minimal filtering approach
print("=" * 60)
print("FILTERING ANALYSIS")
print("=" * 60)

# Define minimal thresholds
MIN_GAMES = 5
MIN_TOTAL_MINUTES = 25

# Show what filtering will do
print(f"\nOriginal dataset: {len(df)} players")
print(f"Filter criteria: GP >= {MIN_GAMES} AND Total Minutes >= {MIN_TOTAL_MINUTES}")

# Apply filter
df_filtered = df[(df['GP'] >= MIN_GAMES) & (df['Total Minutes'] >= MIN_TOTAL_MINUTES)].copy()

removed_count = len(df) - len(df_filtered)
print(f"\n✅ Filtered dataset: {len(df_filtered)} players")
print(f"❌ Removed: {removed_count} players ({removed_count/len(df)*100:.1f}%)")

# Show who was removed (for transparency)
if removed_count > 0:
    removed_players = df[~df.index.isin(df_filtered.index)]
    print(f"\nExamples of removed players (very limited playing time):")
    print(removed_players[['Player Name', 'Team', 'GP', 'Total Minutes', 'MP', 'PTS']].head(10))

FILTERING ANALYSIS

Original dataset: 467 players
Filter criteria: GP >= 5 AND Total Minutes >= 25

✅ Filtered dataset: 436 players
❌ Removed: 31 players (6.6%)

Examples of removed players (very limited playing time):
                 Player Name Team  GP  Total Minutes    MP  PTS
221            Dylan Windler  CLE   3             10   3.3  1.7
247           Sterling Brown  LAL   4             24   6.0  0.0
365          Marko Simonovic  CHI   7             20   2.9  0.9
401  Michael Carter-Williams  ORL   4             44  11.0  4.3
403             Trevor Keels  NYK   3              8   2.7  1.0
404           Dereon Seabron  NOP   5             12   2.4  0.8
406        Mfiondu Kabengele  BOS   4             36   9.0  1.5
408             Malcolm Hill  CHI   5              9   1.8  1.0
413              Darius Days  HOU   4             25   6.3  3.8
428             Chima Moneke  SAC   2              8   4.0  1.0


### Data Quality Verification

After cleaning, the dataset has:
- ✅ **Zero missing values** in all columns
- ✅ **Proper data types** (numeric columns as float64/int64, categorical as object)
- ✅ **467 players** with complete statistical records
- ✅ **51 features** ready for analysis

## Conclusions

**TODO: Write your conclusions here**

Summarize the main insights from your analysis and their implications.

## References

1. Dataset: NBA 2022-23 Season Statistics
2. Libraries used: pandas, numpy, matplotlib, seaborn, plotly, scipy, sklearn
3. [Add any additional references used]